syncstorage-spanner/src/metadata.rs

```diff
@@ -0,0 +1,129 @@
+/// gRPC metadata Resource prefix header
+///
+/// Generic across Google APIs. This "improves routing by the backend" as
+/// described by other clients
+const PREFIX_KEY: &str = "google-cloud-resource-prefix";
+
+/// gRPC metadata Client information header
+///
+/// A `User-Agent` like header, likely its main use is for GCP's metrics
+const METRICS_KEY: &str = "x-goog-api-client";
+
+/// gRPC metadata Dynamic Routing header:
+/// https://google.aip.dev/client-libraries/4222
+///
+/// See the googleapis protobuf for which routing header params are used for
+/// each Spanner operation (under the `google.api.http` option).
+///
+/// https://github.com/googleapis/googleapis/blob/master/google/spanner/v1/spanner.proto
+const ROUTING_KEY: &str = "x-goog-request-params";
+
+/// gRPC metadata Leader Aware Routing header
+///
+/// Not well documented. Added to clients in early 2023 defaulting to disabled.
+/// Clients have began defaulting it to enabled in late 2023.
+///
+/// "Enabling leader aware routing would route all requests in RW/PDML
+/// transactions to the leader region." as described by other Spanner clients
+const LEADER_AWARE_KEY: &str = "x-goog-spanner-route-to-leader";
+
+/// Builds the [grpcio::Metadata] for all db operations
+#[derive(Default)]
+pub struct MetadataBuilder<'a> {
+    prefix: &'a str,
+    routing_params: Vec<(&'a str, &'a str)>,
+    route_to_leader: bool,
+}
+
+impl<'a> MetadataBuilder<'a> {
+    /// Initialize a new builder with a [PREFIX_KEY] header for the given
+    /// resource
+    pub fn with_prefix(prefix: &'a str) -> Self {
+        Self {
+            prefix,
+            ..Default::default()
+        }
+    }
+
+    /// Add a [ROUTING_KEY] header
+    pub fn routing_param(mut self, key: &'a str, value: &'a str) -> Self {
+        self.routing_params.push((key, value));
+        self
+    }
+
+    /// Toggle the [LEADER_AWARE_KEY] header
+    pub fn route_to_leader(mut self, route_to_leader: bool) -> Self {
+        self.route_to_leader = route_to_leader;
+        self
+    }
+
+    /// Build the [grpcio::Metadata]
+    pub fn build(self) -> Result<grpcio::Metadata, grpcio::Error> {
+        let mut meta = grpcio::MetadataBuilder::new();
+        meta.add_str(PREFIX_KEY, self.prefix)?;
+        meta.add_str(METRICS_KEY, "gcp-grpc-rs")?;
```

Comment:
Let's make this a defined constant (so that we can reuse it in the test below.)

Reply:
I'd punt on that for now. Ideally we'd have some kind of more fleshed out "user-agent" like e.g. the python has:

https://github.com/googleapis/python-api-core/blob/main/google/api_core/gapic_v1/client_info.py
https://github.com/googleapis/python-api-core/blob/main/google/api_core/client_info.py

Those are in the more generic library (more along the lines of google-cloud-rust). I'm not sure if individual client libs such as python-spanner further customize these headers.